library/std/src/path.rs

```diff
@@ -979,6 +979,19 @@ impl FusedIterator for Components<'_> {}
 impl<'a> cmp::PartialEq for Components<'a> {
     #[inline]
     fn eq(&self, other: &Components<'a>) -> bool {
+        // fast path for exact matches, e.g. for hashmap lookups
+        if self.path.len() == other.path.len()
+            && self.front == other.front
+            && self.back == State::Body
+            && other.back == State::Body
+        {
+            // this could bail out earlier if there were a reverse memcmp/bcmp comparing back to front
+            if self.path == other.path {
```

Comment:
I'm a little bit wary of this since it doesn't check all the fields -- I suppose we can ignore them because we're already in State::Body (in particular, has_physical_root and prefix)? I think it'd be good to write up a comment explaining why we can do so and potentially adding an explicit de-structuring matching on every field so that we need to edit this code whenever Components changes.

Reply:
> I suppose we can ignore them because we're already in State::Body

We can ignore them because `Eq` only compares the `Component` items, not the prefix. If we're in `front == State::Body` then `prefix` is also covered by `path`. If `Components` has already been partially advanced then the `path` slice doesn't cover it anymore and thus wouldn't be part of the comparison.

But there might be some edge-cases on windows how `CurDir` items are interpreted on verbatim vs. non-verbatim paths so even when the suffix is the same it may have different meaning.

So your concern is warranted. I'll add some extra condition and more comments to check those cases too.

Reply:
I added comments, exra checks and the destructuring. But that destructuring is ugly and overly verbose imo, I would prefer not to have that.

Reply:
Oh, sorry, I think you may have misinterpreted / I was less clear than desirable -- I meant a sort no-op destructuring at the top of the method `let Foo { field1: _, field2: _ } = foo` so that we edit the method if fields are added/removed, but there's no need to make it actually part of the real code.